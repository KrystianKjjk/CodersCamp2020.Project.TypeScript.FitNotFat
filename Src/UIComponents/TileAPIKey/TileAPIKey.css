.api-key-tile {
    background: inherit;
    display: flex;
    flex-direction: column;
    font-weight: 500;
    width: 37rem;
    letter-spacing: 0.035rem;
    margin-left: 1.8rem;
}
.api-key-tile:last-child {
    margin-bottom: 1rem;
}
.api-key-tile__header {
    font-size: 0.8rem;
    color: rgba(74, 72, 78, 0.7);
    background: inherit;
    margin: 1.5rem 0 1rem 0;
}
.api-key-tile__section {
    margin-top: 2rem;
    text-align: center;
}
.api-key-tile__section:nth-of-type(2) {
    margin-bottom: 1.5rem;
}
.api-key-tile__section--text {
    font-size: 0.9rem;
    font-weight: 500;
}
.api-key-tile__section--input {
    text-align: center;
    margin-top: 1.2rem;
    border: #DA1B36 solid 1px;
    outline: none;
    font-size: 1rem;
    height: 2.6rem;
    width: 21rem;
    border-radius: 0.8rem;
    padding: 0.2rem 1rem 0.2rem 1rem;
}
.api-key-tile__section--input:focus {
    border: #DA1B36 solid 2px;
}
