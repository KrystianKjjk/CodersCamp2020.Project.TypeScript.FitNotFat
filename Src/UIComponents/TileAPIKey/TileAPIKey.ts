import { generateWhiteButton } from '../Buttons/Buttons';
import {createElement, createTextInput} from '../utils/utils';
import { DetailsAPI } from "../../../Models/DetailsAPI.model";

const PLACEHOLDER_KEY = 'ex. thisismyapikey';
const PLACEHOLDER_ID = 'ex. thisismyID';

type GetAPIDetails = (username: string) => Promise<DetailsAPI>;
type SetAPIDetails = (username: string, detailsAPI: DetailsAPI) => Promise<string>;

export function createTileAPIKey(username: string, getAPIDetails: GetAPIDetails, setAPIDetails: SetAPIDetails): HTMLElement {

    const containerAPIKey = createElement('div', 'api-key-tile');
    const header = createElement('header', 'api-key-tile__header', 'API KEY');
    const sectionKey = createElement('section', 'api-key-tile__section');
    const sectionKeyText = createElement('p','api-key-tile__section--text', 'ENTER API KEY');
    const sectionKeyInput = createTextInput(PLACEHOLDER_KEY, 'api-key-tile__section--input');
    const sectionID = createElement('section', 'api-key-tile__section');
    const sectionIDText = createElement('p','api-key-tile__section--text', 'ENTER ID');
    const sectionIDInput = createTextInput(PLACEHOLDER_ID, 'api-key-tile__section--input')
    const sectionError = createElement('p','api-key-tile__section--error');
    const saveButton = generateWhiteButton('SAVE', handleSaveButtonClick);
    saveButton.classList.add('my-weight-tile__button-section--tile-btn');

    sectionKey.append(sectionKeyText, sectionKeyInput);
    sectionID.append(sectionIDText, sectionIDInput);
    containerAPIKey.append(header, sectionKey, sectionID, sectionError, saveButton);

    getAPIDetails(username)
        .then((detailsAPI) => {
            sectionKeyInput.value = detailsAPI.key;
            sectionIDInput.value = detailsAPI.id;
        })
        .catch( msg => {});

    function handleSaveButtonClick(){
        const detailsAPI = {
            key: sectionKeyInput.value,
            id: sectionIDInput.value };

        if(detailsAPI.key && detailsAPI.id) {
            setAPIDetails(username, detailsAPI)
                .then( msg => {
                    sectionError.innerText = msg;
                    sectionKeyInput.value = detailsAPI.key;
                    sectionIDInput.value = detailsAPI.id;
                })
                .catch(alert => sectionError.innerText = alert);
        }
        else {
            sectionError.innerText = 'KEY and ID must be filled out!'
        }
    }
    return containerAPIKey;
}
