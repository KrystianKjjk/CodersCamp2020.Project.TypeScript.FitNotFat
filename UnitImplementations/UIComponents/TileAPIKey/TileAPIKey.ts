import { createTileAPIKey } from '../../../Src/UIComponents/TileAPIKey/TileAPIKey';
import  generateTileComponent  from '../../../Src/UIComponents/TileComponent/TileComponent'
import { DetailsAPI } from "../../../Models/DetailsAPI.model";
import { User } from "../../../Models/User.model";


const username: string = 'User';
type GetAPIDetails = (username: string) => Promise<DetailsAPI>;
type SetAPIDetails = (username: string, detailsAPI: DetailsAPI) => Promise<string>;

let getAPIDetails: GetAPIDetails = function(username) {
    return new Promise( (resolve, reject)=>{reject(new Error(`User doesn't exist!`));} );
}
let setAPIDetails: SetAPIDetails = function(username, detailsAPI) {
    return new Promise( (resolve, reject)=>{ resolve('API DETAILS saved correctly!');} );
}

const tileMyWeight = generateTileComponent(
    createTileAPIKey(username, getAPIDetails, setAPIDetails));

document.body.appendChild(tileMyWeight);


