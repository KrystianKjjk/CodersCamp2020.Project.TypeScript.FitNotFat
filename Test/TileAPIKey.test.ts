import { createTileAPIKey } from "../Src/UIComponents/TileAPIKey/TileAPIKey";
import {DetailsAPI} from "../Models/DetailsAPI.model";
import generateForm from "../Src/UIComponents/FormComponent/FormComponent";


describe('TileAPIKey component tests', ()=> {

    const PLACEHOLDER_KEY = 'ex. thisismyapikey';
    const PLACEHOLDER_ID = 'ex. thisismyID';
    const username = 'User'
    const getAPI = jest.fn(username => Promise.resolve({key: '', id: ''}));
    const setAPI = jest.fn(username => Promise.reject(''));

    test('if returns HTMLElement', () => {
        expect(createTileAPIKey(username, getAPI, setAPI)).toBeInstanceOf(HTMLElement);
    })
    test('if renders properly',()=>{
        const apiKey = createTileAPIKey(username, getAPI, setAPI);
        expect(apiKey).toMatchSnapshot();
    })
    test('if contains default placeholder', () => {
        let element = createTileAPIKey(username, getAPI, setAPI);
        expect((element.querySelectorAll('.api-key-tile__section--input')[0] as HTMLInputElement).placeholder).toBe(PLACEHOLDER_KEY);
        expect((element.querySelectorAll('.api-key-tile__section--input')[1] as HTMLInputElement).placeholder).toBe(PLACEHOLDER_ID);
    })
    test('if contains new values in inputs', async () => {
        const KEY_VALUE = 'newapiKEY';
        const ID_VALUE = 'newapiID';

        const getAPI = jest.fn(username => Promise.resolve({key: KEY_VALUE, id: ID_VALUE}));
        let element = await createTileAPIKey(username, getAPI, setAPI);
        expect((element.querySelectorAll('.api-key-tile__section--input')[0] as HTMLInputElement).value).toBe(KEY_VALUE);
        expect((element.querySelectorAll('.api-key-tile__section--input')[1] as HTMLInputElement).value).toBe(ID_VALUE);
    })
    test('if clicking button when inputs are empty displays information', async () => {
        const element = createTileAPIKey(username, getAPI, setAPI);

        (element.querySelector('.my-weight-tile__button-section--tile-btn') as HTMLButtonElement).click();
        expect((element.querySelector('.api-key-tile__section--error') as HTMLElement).innerText).toBe('KEY and ID must be filled out!');
    })
    test('if clicking save button when inputs are filled up updates values in input fields', async () => {
        const KEY_VALUE = 'newapiKEY';
        const ID_VALUE = 'newapiID';

        window.alert = jest.fn();
        const element = createTileAPIKey(username, getAPI, setAPI);
        (element.querySelectorAll('.api-key-tile__section--input')[0] as HTMLInputElement).value = KEY_VALUE;
        (element.querySelectorAll('.api-key-tile__section--input')[1] as HTMLInputElement).value = ID_VALUE;
        (element.querySelector('.my-weight-tile__button-section--tile-btn') as HTMLButtonElement).click();
        expect((element.querySelectorAll('.api-key-tile__section--input')[0] as HTMLInputElement).value).toBe(KEY_VALUE);
        expect((element.querySelectorAll('.api-key-tile__section--input')[1] as HTMLInputElement).value).toBe(ID_VALUE);
    })
});
