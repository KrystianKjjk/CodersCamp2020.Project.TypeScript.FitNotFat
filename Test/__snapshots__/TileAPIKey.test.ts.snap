// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TileAPIKey component tests if renders properly 1`] = `
<div
  class="api-key-tile"
>
  <header
    class="api-key-tile__header"
  >
    API KEY
  </header>
  <section
    class="api-key-tile__section"
  >
    <p
      class="api-key-tile__section--text"
    >
      ENTER API KEY
    </p>
    <input
      class="api-key-tile__section--input"
      placeholder="ex. thisismyapikey"
      required=""
      type="text"
    />
  </section>
  <section
    class="api-key-tile__section"
  >
    <p
      class="api-key-tile__section--text"
    >
      ENTER ID
    </p>
    <input
      class="api-key-tile__section--input"
      placeholder="ex. thisismyID"
      required=""
      type="text"
    />
  </section>
  <p
    class="api-key-tile__section--error"
  />
  <button
    class="button my-weight-tile__button-section--tile-btn"
    style="background-color: white; color: rgb(218, 27, 54);"
    type="button"
  />
</div>
`;
